Fix memory leak in TensorPipeAgent.

pritamdamania87 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#50564 Fix memory leak in TensorPipeAgent.**

When an RPC was sent, the associated future was stored in two maps:
pendingResponseMessage_ and timeoutMap_. Once the response was received, the
entry was only removed from pendingResponseMessage_ and not timeoutMap_. The
pollTimedoudRpcs method then eventually removed the entry from timeoutMap_
after the time out duration had passed.

Although, in scenarios where there is a large timeout and a large number of
RPCs being used, it is very easy for the timeoutMap_ to grow without any
bounds. This was discovered in https://github.com/pytorch/pytorch/issues/50522.

To fix this issue, I've added some code to cleanup timeoutMap_ as well once we
receive a response.

#Closes: https://github.com/pytorch/pytorch/issues/50522

Differential Revision: [D25919650](https://our.internmc.facebook.com/intern/diff/D25919650/)

pritamdamania committed:
Fix memory leak in TensorPipeAgent.

When an RPC was sent, the associated future was stored in two maps:
pendingResponseMessage_ and timeoutMap_. Once the response was received, the
entry was only removed from pendingResponseMessage_ and not timeoutMap_. The
pollTimedoudRpcs method then eventually removed the entry from timeoutMap_
after the time out duration had passed.

Although, in scenarios where there is a large timeout and a large number of
RPCs being used, it is very easy for the timeoutMap_ to grow without any
bounds. This was discovered in https://github.com/pytorch/pytorch/issues/50522.

To fix this issue, I've added some code to cleanup timeoutMap_ as well once we
receive a response.

Differential Revision: [D25919650](https://our.internmc.facebook.com/intern/diff/D25919650/)

[ghstack-poisoned]

rohan-varma commented on torch/csrc/distributed/rpc/tensorpipe_agent.cpp:
```diff
@@ -768,6 +770,28 @@ std::shared_ptr<JitFuture> TensorPipeAgent::send(
+                for (auto it = timedOutFuturesVector.begin();
+                     it != timedOutFuturesVector.end();) {
+                  if (std::get<0>(*it) == messageId) {
+                    it = timedOutFuturesVector.erase(it);
```
We can `break` here since messageId should be unique right?

mrshenli commented on torch/csrc/distributed/rpc/tensorpipe_agent.cpp:
```diff
@@ -678,7 +678,9 @@ std::shared_ptr<JitFuture> TensorPipeAgent::send(
-      timeoutFuturesVector.emplace_back(futureResponseMessage, timeout);
+      timeoutFuturesVector.emplace_back(
+          messageId, futureResponseMessage, timeout);
+      messageIdToExpiryMap_[messageId] = expirationTime;
```
nit, feel free to ignore, if all we need is to pass the `expirationTime` to the response processing logic, we can potentially capture it in `pipeWrite` and then `pipeRead` to avoid global map insertion and lookup?

mrshenli commented on torch/csrc/distributed/rpc/tensorpipe_agent.h:
```diff
@@ -2,6 +2,7 @@
 
 #ifdef USE_TENSORPIPE
 
+#include <gtest/gtest_prod.h>
```
cc @lw to check if this is OK.

lw replied:
First time I see this, I assume it's for the `FRIEND_TEST` macro. In that case it looks fine to me. My only concern is whether we're sure that gtest is always available at build time? (Maybe there's a "no-test" build mode where gtest could be missing?)

mrshenli reviewed (approved):
LGTM! Stamp to unblock. Could you please verify that this change does not cause perf regression using our `test_stress_*`? Thanks!

rohan-varma commented on torch/csrc/distributed/rpc/tensorpipe_agent.h:
```diff
@@ -319,11 +324,16 @@ class TensorPipeAgent : public RpcAgent {
       steady_clock_time_point,
-      std::vector<std::pair<
+      std::vector<std::tuple<
+          uint64_t, // messageId
```
Nit, and unrelated to this diff: Now that we have messageId here, we can probably remove the `std::shared_ptr<AtomicJitFuture>` key from this map and just use the messageId to key into the pending futures map. This is similar to what process group agent does.

lw commented on torch/csrc/distributed/rpc/tensorpipe_agent.cpp:
```diff
@@ -768,6 +770,28 @@ std::shared_ptr<JitFuture> TensorPipeAgent::send(
                 clientPipe.pendingResponseMessage_.erase(it);
               }
 
+              // Remove entry from timeoutMap_ as well.
```
We might want to do the same in the error branch, i.e., at lines 743-745. What happens there is that all pending messages for this outgoing pipe are instantly completed with error, and dropped from `clientPipe.pendingResponseMessage_`. To make it neater we could also remove them from the `messageIdToExpiryMap_` and the `timeoutMap_`.

lw commented on torch/csrc/distributed/rpc/tensorpipe_agent.h:
```diff
@@ -319,11 +324,16 @@ class TensorPipeAgent : public RpcAgent {
       steady_clock_time_point,
-      std::vector<std::pair<
+      std::vector<std::tuple<
+          uint64_t, // messageId
           std::shared_ptr<AtomicJitFuture>,
-          std::chrono::milliseconds>>>
+          std::chrono::milliseconds>>> // timeout
```
Nit: for readability it may make sense to define a custom struct instead of this tuple, so that fields can be accessed with meaningful names, rather than with `std::get`.

pritamdamania committed:
Update on "Fix memory leak in TensorPipeAgent."


When an RPC was sent, the associated future was stored in two maps:
pendingResponseMessage_ and timeoutMap_. Once the response was received, the
entry was only removed from pendingResponseMessage_ and not timeoutMap_. The
pollTimedoudRpcs method then eventually removed the entry from timeoutMap_
after the time out duration had passed.

Although, in scenarios where there is a large timeout and a large number of
RPCs being used, it is very easy for the timeoutMap_ to grow without any
bounds. This was discovered in https://github.com/pytorch/pytorch/issues/50522.

To fix this issue, I've added some code to cleanup timeoutMap_ as well once we
receive a response.

#Closes: https://github.com/pytorch/pytorch/issues/50522

Differential Revision: [D25919650](https://our.internmc.facebook.com/intern/diff/D25919650/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Fix memory leak in TensorPipeAgent."


When an RPC was sent, the associated future was stored in two maps:
pendingResponseMessage_ and timeoutMap_. Once the response was received, the
entry was only removed from pendingResponseMessage_ and not timeoutMap_. The
pollTimedoudRpcs method then eventually removed the entry from timeoutMap_
after the time out duration had passed.

Although, in scenarios where there is a large timeout and a large number of
RPCs being used, it is very easy for the timeoutMap_ to grow without any
bounds. This was discovered in https://github.com/pytorch/pytorch/issues/50522.

To fix this issue, I've added some code to cleanup timeoutMap_ as well once we
receive a response.

#Closes: https://github.com/pytorch/pytorch/issues/50522

Differential Revision: [D25919650](https://our.internmc.facebook.com/intern/diff/D25919650/)

[ghstack-poisoned]
